### ISSUE TITLE
[Performance] vLLM warmup does not cover C128A extra_topk=384 on SM120 sparse-MLA decode, causing SparseMlaDecodeV3Runner tactic=-1 fallback

### ISSUE THREAD

**Issue by wuhangxian** [vllm-project/vllm#53600]
## Summary

vLLM startup warmup for DeepSeek-V4-Flash with SM120 (RTX PRO 5000 Blackwell) sparse-MLA decode only generates FlashInfer autotune cache entries for `extra_topk={0, 128, 512}`. At runtime, C128A compressed cache dynamically selects `extra_topk=256` (seq 16,512–32,895) or `extra_topk=384` (seq 32,896–33,792). Since both widths are missing from the autotune cache, every long-sequence decode step falls back to `SparseMlaDecodeV3Runner tactic=-1`, degrading performance.

This is a **performance** issue (not a correctness bug): `tactic=-1` is FlashInfer's legitimate heuristic fallback kernel that produces mathematically correct results, just with a suboptimal tactic. The issue is classified as performance because the functional output is correct; only the kernel selection is suboptimal.

### Hardware

| Item | Value |
| --- | --- |
| GPU | 4 × NVIDIA RTX PRO 5000 72GB Blackwell (SM120), TP4 |
| NVIDIA Driver | 580.126.20 |

### Software

| Item | Value |
| --- | --- |
| vLLM | `0.1.dev20005+g75c2eef60` (custom build based on current main) |
| flashinfer-python | `0.6.18.dev20260811` |
| flashinfer-cubin | `0.6.18.dev20260811` |
| flashinfer-jit-cache | `0.6.18.dev20260811+cu130` |
| PyTorch | `2.13.0+cu130` |
| CUDA | 13.0 |

### Container image

The reproduction uses a custom image built from the vLLM cu13 nightly base, with FlashInfer upgraded from 0.6.16.post3 to 0.6.18.dev20260811.

DeepSeek-V4-Flash requires FlashInfer 0.6.18.dev because the base nightly image ships FlashInfer 0.6.16.post3, whose SM120 sparse-MLA prefill kernel lacks the `num_heads=8` branch — at TP=8 (8 heads/card), this crashes on startup. FlashInfer 0.6.18.dev (PR [#4380](https://github.com/flashinfer-ai/flashinfer/pull/4380), merged 2026-08-08) adds the NH=8 padded-tile prefill kernel and fixes DSpark draft-decode token routing, making DeepSeek-V4-Flash bootable on SM120.

**Image pull command:**
```bash
docker pull hai-beijing.tencentcloudcr.com/dorianwu-sglang/vllm:hai-dsv4flash-pro5000
```

This image only modifies the three FlashInfer wheels (python/cubin/jit_cache); nothing else is changed vs the base `hai-beijing.tencentcloudcr.com/dorianwu-sglang/vllm:nightly-dev-cu13` (vLLM commit `75c2eef`, FlashInfer 0.6.16.post3). The image used in this report is tagged `hai-deepseek-v4-flash-vllm:latest` (`sha256:113c108a11f3...4086cee`), which is a local re-tag of the same image.

### Model configuration

| Item | Value |
| --- | --- |
| Model | DeepSeek-V4-Flash (FP8 KV cache, speculative with DSpark, 5 spec tokens) |
| max_model_len | 33,792 |
| max_num_batched_tokens | 8,448 |
| max_num_seqs | 32 |
| tensor_parallel_size | 4 |
| enable_flashinfer_autotune | True (default for this model) |
| enable_prefix_caching | False |

### Launch command

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 \\
python3 -m vllm.entrypoints.openai.api_server \\
  --model /model \\
  --host 0.0.0.0 \\
  --port 30013 \\
  --trust-remote-code \\
  --served-model-name DeepSeek-V4-Flash \\
  --kv-cache-dtype fp8 \\
  --gpu-memory-utilization 0.8 \\
  --speculative-config \'{"method":"dspark","num_speculative_tokens":5}\' \\
  --max-model-len 33792 \\
  --max-num-batched-tokens 8448 \\
  --max-num-seqs 32 \\
  --tensor-parallel-size 4 \\
  --seed 0 \\
  --no-enable-prefix-caching
```

### Observed behavior during startup

During warmup, the log shows:
```
flashinfer_sparse_mla_warmup.py:233 - Warming up DeepSeek V4 sparse MLA attention for mixed tokens=16.
flashinfer_sparse_mla_warmup.py:124 - Autotuning FlashInfer SM120 sparse MLA DSv4 decode with cache: .../autotune_configs.json
```

The autotune warmup generates **24 cache entries** with `extra_topk ∈ {0, 128, 512}` only — **zero entries for `extra_topk=256` and `extra_topk=384`**.

During CUDA Graph capture, **944 `tactic=-1` fallback warnings** are emitted across all 4 TP workers, covering 4 unique batch shapes (all with `extra_topk=384`):

| Batch (T) | extra_topk | T bucket | Warnings per TP | Total (×4 TP) |
| ---: | ---: | ---: | ---: | ---: |
| 6 | 384 | 8 | 236 | 236 |
| 12 | 384 | 16 | 236 | 236 |
| 24 | 384 | 32 | 236 | 236 |
| 60 | 384 | 64 | 236 | 236 |

**Total: 944 warnings** (4 unique shapes × 236 per shape × 1 per TP worker per shape = 944).

Note: 256-width fallbacks were not observed in this specific run because the CUDA Graph capture sizes (`[1, 2, 4, 8, 16, 24, 32, ..., 384]`) with `max_model_len=33792` and DSpark 5 spec tokens produce `extra_topk=384` at all capture sizes. However, `extra_topk=256` is statically reachable (seq 16,512–32,895) and would miss at runtime just as 384 does.

Example warning:
```
WARNING - autotuner.py:1513 - flashinfer.jit: [AutoTuner]: No tuned config covers
sparse_mla_sm120_decode_dsv4 input_shapes=((60, 16, 512), (60, 128), (60, 16, 8, 512),
(60, 16, 8), (60, 16, 512), (60, 16), (60,), (16,), (60, 384), (60,));
falling back to runner=SparseMlaDecodeV3Runner tactic=-1.  This shape is outside the
tuning bucket range -- expand tuning_buckets / max_num_tokens during the next tuning
pass to avoid this perf cliff.
```

### Cache analysis

The autotune cache file at:
`/root/.cache/vllm/flashinfer_autotune_cache/0.6.18.dev20260811/120f/96af87eb6dbf9df352c4360d1a159c7bf53718f047c06b072979c73c88c0965e/autotune_configs.json`

contains 24 entries, all with `extra_topk ∈ {0, 128, 512}`:

| extra_topk | Records per T bucket {1,4,8,16,32,64} | Total |
| ---: | ---: | ---: |
| 0 | 1 each | 6 |
| 128 | 1 each | 6 |
| 512 | 1 each (varying tactics) | 6 |
| **256** | **0** | **0** |
| **384** | **0** | **0** |

Cache SHA-256: `38bf5e3d0b960250eff60e9ea5ea77fe7d29b989c88f1d3313cc6daadd805265`

### Fail-closed request validation

Three requests were sent to the running server. All completed with exact requested output token counts. The cache file was unchanged before and after (SHA-256 verified), confirming that runtime requests do not auto-populate missing cache entries:

| Input tokens | Output tokens | Elapsed | Finish reason |
| ---: | ---: | ---: | --- |
| 8193 | 512 | 1.92 s | length |
| 16513 | 64 | 1.59 s | length |
| 32897 | 64 | 3.13 s | length |

Cache SHA-256 before: `38bf5e3d0b960250eff60e9ea5ea77fe7d29b989c88f1d3313cc6daadd805265`
Cache SHA-256 after:  `38bf5e3d0b960250eff60e9ea5ea77fe7d29b989c88f1d3313cc6daadd805265`
Cache unchanged: **Yes**

### Background: What is C128A and extra_topk?

DeepSeek-V4-Flash uses a C128A compressed KV cache: every 128 consecutive tokens are compressed into 1 "summary" token. During decode, the attention layer selects the most relevant compressed summaries — the number selected is called `extra_topk`. This value scales with the request's sequence length: longer sequences need more summaries to cover the full context.

The `extra_topk` value is not arbitrary — it is computed as `next_power_of_2(max_seq_len // 128)`, clamped to a minimum of 128 (the C128A tile alignment) and a maximum of `c128a_max_compressed` (the total number of compressed tokens for `max_model_len`). For `max_model_len=33,792`, this maximum is `384`.

### 1. C128A runtime width dynamically ranges over {128, 256, 384}

In `vllm/models/deepseek_v4/sparse_mla.py`, the C128A compressed cache width is computed at initialization:

```python
# sparse_mla.py:166-170 (initialization)
c128a_max_compressed = cdiv(max_model_len, compress_ratio)       # cdiv(33792, 128) = 264
c128a_max_compressed = cdiv(c128a_max_compressed, _C128A_TOPK_ALIGNMENT) * _C128A_TOPK_ALIGNMENT
                                                                    # cdiv(264, 128) * 128 = 384
```

At runtime, the active width is dynamically selected based on the current batch's `max_seq_len`:

```python
# sparse_mla.py:261-265 (per-batch, at metadata build time)
active_topk_width = min(
    max(
        triton.next_power_of_2(max(cm.max_seq_len // self.compress_ratio, 1)),
        _C128A_TOPK_ALIGNMENT,   # 128
    ),
    self.c128a_max_compressed,    # 384
)
```

For `max_model_len=33,792`, the three runtime-reachable widths are:

| Request max_seq_len | max_seq_len // 128 | next_power_of_2 | active_topk_width |
| ---: | ---: | ---: | ---: |
| 1–16,511 | 0–128 | 1–128 | **128** |
| 16,512–32,895 | 129–256 | 256 | **256** |
| 32,896–33,792 | 257–264 | 512 → clamped | **384** |

All three widths are statically reachable in production. Any request with context length ≥ 16,512 tokens will miss the cache.

### 2. FlashInfer autotune cache keys on exact `extra_topk`

FlashInfer's sparse-MLA autotune cache uses the exact `extra_topk` value as part of the cache key. A tactic tuned for `extra_topk=512` cannot cover `extra_topk=384` — they are separate entries. The T (batch size) dimension uses ceiling buckets `{1, 4, 8, 16, 32, 64}`, so a T=60 request maps to bucket 64. The cache miss is specifically on the `extra_topk` width dimension, not the T dimension.

### 3. Warmup only exercises `extra_topk=128`

The warmup path in `vllm/model_executor/warmup/flashinfer_sparse_mla_warmup.py` uses:

```python
_SPARSE_MLA_MIXED_WARMUP_TOKENS = 16
```

This creates a mixed batch (`create_mixed_batch=True`) in `_dummy_run` where decode tokens have `seq_lens=[1]` (`gpu_model_runner.py:6131`):

```python
# gpu_model_runner.py:6131 (mixed batch for warmup)
seq_lens = torch.tensor([1] * num_decode_tokens + [num_prefill_tokens + 1], ...)
```

With `seq_lens=1`, the C128A width computation yields:
```
active_topk_width = min(max(next_power_of_2(1 // 128), 128), 384) = 128
```

The warmup **never simulates a decode request with `seq_len ≥ 16,512`**, so `extra_topk=256` and `extra_topk=384` are never autotuned. The 16-token mixed batch is sufficient for JIT compilation warmup (triggering kernel compilation and loading), but it does not cover the autotune shape space for C128A widths.

### Why the warmup gap was not caught earlier

The warmup code was designed for JIT compilation warmup — ensuring that kernels are compiled and loaded before serving requests. The `_dummy_run` with 16 tokens achieves this goal. However, the same `_dummy_run` is also used for FlashInfer autotuning (wrapped in `flashinfer_autotune(True, ...)` context), which requires shape coverage, not just compilation. The two concerns (JIT warmup vs. autotune shape coverage) were conflated into a single dummy run with a minimal shape, leaving the autotune cache incomplete for C128A widths beyond 128.

## Expected behavior

- Startup autotune covers **every** runtime-reachable C128A width: `{128, 256, 384}` for `max_model_len=33,792`.
- A production shape does not use `tactic=-1` only because warmup selected a different sequence-length bucket.
- Existing cache entries (for `extra_topk=0, 128, 512`) survive the additional warmup passes.
- Partial tuning failure preserves the baseline cache and cannot deadlock TP followers.
- Repeated startup with a complete cache is idempotent.

## Proposed fix

The warmup should enumerate all runtime-reachable C128A widths and exercise each through the autotune path. For each width, a `_dummy_run` with `profile_seq_lens` set to a representative `seq_len` that triggers that width should be run under the `flashinfer_autotune` context.

Concretely, the warmup should run additional autotune passes with:
- `profile_seq_lens=16512` → triggers `extra_topk=256`
- `profile_seq_lens=max_model_len` → triggers `extra_topk=384`

The existing pass (`seq_lens=[1]`) already covers `extra_topk=128`.

See linked PR #53605 for implementation.

## Additional context

- The same defect was independently reproduced on vLLM current main (`a047e2543`) with FlashInfer 0.6.17 (pinned dependency), confirming it is not version-specific to 0.6.18.dev.
- In the motivating environment, adding real width-384 tactics (manual cache intervention) removed all 944 fallback warnings but did not make target logits bit-reproducible.

## Before submitting a new issue

- [x] Searched open and closed vLLM issues and PRs for `extra_topk`, `384`, `sparse MLA`, `FlashInfer`, `autotune`, and `warmup`.
- [x] Found related reports, but no issue or PR covering the missing C128A production-width tuning keys.
- [x] Reproduced the scoped width-384 defect on the production environment (FlashInfer 0.6.18.dev20260811) and independently on current main + FlashInfer 0.6.17.
